docs/proposal.md

```diff
@@ -0,0 +1,108 @@
+CRI-containerd relies on containerd to manage container lifecycle.
+
+Ideally, CRI-containerd only needs to do api translation and information reorganization. However, CRI-containerd needs to maintain some metadata because:
+* There is a mismatch between container state machine of containerd and CRI - containerd doesn't support `Stop`, so CRI-containerd needs to maintain the `Stopped` state.
+* Some sandbox/container metadata is not provided by containerd, and we can not leverage OCI runtime annotation to store it because of the state machine mismatch, e.g. labels/annotations, `PodSandboxID` of a container, `FinishedAt` timestamp, `ExitCode`, `Mounts` etc.
+
+CRI-containerd should checkpoint these metadata itself or use [containerd metadata service](https://github.com/containerd/containerd/blob/0a5544d8c4dab44dfc682f5ad07f1cd011c0a115/design/plugins.md#core) if available.
+### Container Logging
+Containerd doesn't provide persistent container log. It redirects container STDIO into different FIFOs.
+
+CRI-containerd should start a goroutine (process/container in the future) to:
+* Continuously drain the FIFO;
+* Decorate the log line into [CRI-defined format](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/kubelet-cri-logging.md#proposed-solution);
+* Write the log into [CRI-defined log path](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/kubelet-cri-logging.md#proposed-solution).
+### Container Streaming
+Containerd supports creating a process in the container with `Exec`, and the STDIO is also exposed as FIFOs. Containerd also supports resizing console of a specific process with `Pty`.
+
+CRI-containerd could reuse the [streaming server](https://github.com/kubernetes/kubernetes/blob/release-1.6/pkg/kubelet/server/streaming/server.go), it should implement the [streaming runtime interface](https://github.com/kubernetes/kubernetes/blob/release-1.6/pkg/kubelet/server/streaming/server.go#L61-L65).
+
+For different CRI streaming functions:
+* `ExecSync`: CRI-containerd should use `Exec` to create the exec process, collect the stdout/stderr of the process, and wait for the process to terminate.
+* `Exec`: CRI-containerd should use `Exec` to create the exec process, create a goroutine (process/container) to redirect streams, and wait for the process to terminate.
+* `Attach`: CRI-containerd should create a goroutine (process/container) to read the existing container log to the output, redirect streams of the init process, and wait for any stream to be closed.
+* `PortForward`: CRI-containerd could implement this with `socat` and `nsenter`, similar with [current Docker portforward implementation](https://github.com/kubernetes/kubernetes/blob/release-1.6/pkg/kubelet/dockertools/docker_manager.go#L1373-L1428).
+### Container Networking
+Containerd doesn't provide container networking, but OCI runtime spec supports joining a container into an existing network namespace.
```

Comment:
> /s/provide/currently provide/

They won't provide container networking, even in the future. And that's what we want.

> /s/OCI runtime spec/the current OCI runtime spec, as implemented by containerd via runc,/

Done.